Change user and database creation method to salt module states

markuskurbel opened the pull request:
A change of user and database creation to the salt module states. This is a major update.
Note: Due to the new db creation method, the ownership behaviour changed to: last user will be the owner of the created db. This also updates now, in case you append the list! This is potentially confusing.

markuskurbel committed:
Change user and database creation method to salt module states

chr4 commented on init.sls:
```diff
@@ -106,20 +106,48 @@ chown_pgdata:
+    - name: {{ config['username'] }}
+    - encrypted: True
+    {% if config['login']|default(false) %}
+    - login: True
```
I think you could use the pillar value here directly:

```
- login: {{ config['login'] | default(false) %}
```

chr4 replied:
Also we should adhere to the defaults. Login for example defaults to `true`. 
See: https://www.postgresql.org/docs/11/app-createuser.html

markuskurbel replied:
As I wrote in the initial comment, the ownership behavior is changeing from first defined user, to last defined user, which even updates on append. This is kind of counter intuitiv, right? I am not sure if we should use the ```postgres_database.present``` state in the first place. WHat do you think?

chr4 replied:
There's a Jinja filter called [reverse()](https://jinja.palletsprojects.com/en/2.10.x/templates/#reverse) maybe just use that in the for loop?

markuskurbel replied:
Yeah, it is kind of irritating, but the ```postgres_user``` state is useing ```CREATE ROLE```, which defaults to NOLOGIN.
https://github.com/saltstack/salt/blob/develop/salt/modules/postgres.py#L1149
https://www.postgresql.org/docs/current/sql-createrole.html

chr4 replied:
I agree that this is confusing.
I think we should default to `true` here, because our pillar entry is called `postgresq:users` and not `postgresql:roles`, and I think we should therefore use the default for `CREATE USER`, at least to me this seems more intuitively.

Alternatively, we could rename everything to `roles`, which might be the more correct approach, but I think we should hold that back for a future release (maybe), what do you think?

markuskurbel replied:
Yes i agree, that it is more intuitively and more consistent with the strange nameing of the salt state. Basically we are creating an ```alias```, as it is done in ```CREATE USER```.
I guess this is also better, because ```login: true``` is probably the more frequently used option.

chr4 commented on init.sls:
```diff
@@ -106,20 +106,48 @@ chown_pgdata:
+    - password: {{ config['password'] }}
+    {% endif %}
+    {% if config['createdb']|default(false) %}
+    - createdb: True
```
Same for the other boolean args

chr4 commented on init.sls:
```diff
@@ -106,20 +106,48 @@ chown_pgdata:
+    - createdb: True
+    {% endif %}
+    {% if config['inherit']|default(false) %}
+    - inherit: True
```
+1

chr4 commented on init.sls:
```diff
@@ -106,20 +106,48 @@ chown_pgdata:
+    - inherit: True
+    {% endif %}
+    {% if config['replication']|default(false) %}
+    - replication: True
```
+1

chr4 commented on pillar.example:
```diff
@@ -76,27 +76,33 @@ postgresql:
       type: host
       database: production
-      address: 10.1.2.0/24
+      address: 0.0.0.0/24
```
`0.0.0.0/24` makes no sense? I think we should continue using some example address here. Or was there a special reason you've changed it?

markuskurbel replied:
Just for testing. I will change it back

chr4 commented on init.sls:
```diff
@@ -106,20 +106,48 @@ chown_pgdata:
+    {% endif %}
+    {% if config['replication']|default(false) %}
+    - replication: True
+    {% endif %}
```
I think we should also support the other [options](https://docs.saltstack.com/en/latest/ref/states/all/salt.states.postgres_user.html): superuser, createroles, createdb

chr4 commented:
This looks good. Let me test-apply this to some systems for further testing. Bear with me for a while.

markuskurbel commented:
Sorry, I just changed the pillar to reflect our new login default value. ```login: true``` removed, ```login: false``` added

chr4 commented on init.sls:
```diff
@@ -104,22 +104,45 @@ chown_pgdata:
 
 # Deploy users. Sort them, so order is not changing between salt runs
-{% for config in pillar['postgresql']['users']|default({}) %}
+# Reverse the order for more convenient ownership of databases in multiuser scenarios
```
I think there should be a not on what's meat by this. Something like "Because `postgres_user.present` automatically promotes the last user as the database owner, we're reversing the order. This ensures backwards compatibility to previous configurations and also is more intuitive" - or something like that.

markuskurbel replied:
What about:
Reverse the order so the first user defined in the pillar is database owner (as in previous versions), since postgres_database.present updates ownership in every iteration

chr4 replied:
perfect!

chr4 commented on pillar.example:
```diff
@@ -79,24 +79,30 @@ postgresql:
-      options: NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT LOGIN REPLICATION
-      password:
+      method: md5
+      password: password
```
theres `replication: true` missing for this example.

chr4 commented:
There's one small adaption needed: `inherit` is also enabled by default and should be `true` by default as well.
Besides that, it looks very good!

Also, the options should be mentioned in the README, maybe just add all of them to the first user created with their respective default values (and add a comment that the examples show the default values): 

```
superuser
createdb
createroles
inherit
replication 
```

chr4 commented:
Also, as this is a major release, we should introduce a CHANGELOG that describes the changes to `OPTIONS` and points to the new pillar parameters instead.

chr4 commented:
This can be done outside of this PR though.

markuskurbel committed:
Integrate review comments

- Add more pillar entries for documentation
- Change inherit option to postgres default
- Remove reverse function from database creation state

markuskurbel commented:
There is actually a little bug/inconvenience in this. When you create a user with role, before the role is created, it obviously breaks. Applying the state twice works.  
However I am not sure if this is also worth documenting somewhere.

chr4 commented on pillar.example:
```diff
@@ -84,6 +84,12 @@ postgresql:
       database: production
       address: 10.1.2.0/24
       method: trust
+      #NOTE: The following values represent the defaults and are for documentation purposes only.
```
Please add a space after the `#`.
I think the "documentation only" section can be removed, as the values are completely valid. 

```
# The following values represent the defaults
```

Would be sufficient.

chr4 commented:
I'm not sure I understand. Can you provide an example YAML?

markuskurbel commented:
Actually, I just realised, the example pillar has this flaw. As we used the ```reverse``` in the state we apply ```example_user_with_role``` before ```example_role```. Hence the ```example_user_with_role``` creation will fail, because its role does not yet.

chr4 commented:
Ah ok. So `groups` actually refers to `roles`? That's weird - but I guess ok for us, as we're just passing through the module parameters. 

That's super unintuitive, so maybe we should cancel the `reverse` and add a respective note to the examples (that the last user is going to become owner)?

Can you explain the initial problem again, what you meant with "last user will be the owner of the created db"? The `database` section has an `owner` attribute, so it should be set explicitly? OR does this occur when using `createdb` with a user, but somehow assigning more users to it (not sure if that's a good practise)?

markuskurbel commented:
Whenever you want to create a database with multiple users the last user defined for that database will be the owner, because the ```postgres_database.present``` state updates it on every apply.  
I find this behaviour inconvenient, because intuitively I would append a list, in order to add something. However this results in a change of ownership.  
I am not sure what you consider bad practice, but i mean adding a new user and create the appropriate ```pg_hba``` entries.

chr4 commented:
But [postgres_database.present](https://docs.saltstack.com/en/latest/ref/states/all/salt.states.postgres_database.html) has an `owner` attribute, where this could just be set manually if needed? 

We could just pass that through and only default to `config['username']` if the `owner` field is not set explicitly? That would be a clean approach I feel. In both cases, we should remove the `reverse()` filter.

markuskurbel commented:
Yes, I also thought about this. However, then we would set an ```owner``` within an ```user``` context, which "feels" wrong. Consider we are doing this, it would mean that for the owner, no ```owner``` value is needed. For all other users this needs to be set. I guess this behaviour needs to be commented as well, because if you don't follow this pattern, the user gets the strange ownership changes described above.

markuskurbel commented:
Maybe it is also worth considering, rolling back the DB creation to plain SQL again.

chr4 commented:
Or we introduce a proper `database` configuration section, analogue to users, but for databases.

markuskurbel commented:
Yes, also possible. I guess it is best to have a chat about this at some point, to align.
